### ISSUE TITLE
rrule _until missing timezone info

### ISSUE THREAD

**Issue by jeinarsson** [dateutil/dateutil#620]
I have a calendar (generated by Google) with a line like so:

> RRULE:FREQ=MONTHLY;UNTIL=20180819;INTERVAL=12;BYDAY=3MO

If I use `rrulestr` to parse this recurrence with a `dtstart` **including timezone info**, the generator fails, because it tries a comparison with the `_until` field which has no timezone info.

Here's a minimal example of what I mean:

```python
from datetime import datetime, timezone
from dateutil.rrule import *

rule_str = 'FREQ=MONTHLY;UNTIL=20180819;INTERVAL=12;BYDAY=3MO'

# this works (no timezone data)
start = datetime(2017,8,21)
rule = rrulestr(rule_str, dtstart=start)
print(list(rule))

# but this does not
start = datetime(2017,8,21,tzinfo=timezone.utc)
rule = rrulestr(rule_str, dtstart=start)
print(list(rule))
```

Output:

```
[datetime.datetime(2017, 8, 21, 0, 0)]
Traceback (most recent call last):
  File "mwe_rrule.py", line 17, in <module>
    print(list(rule))
  File "/usr/local/lib/python3.6/site-packages/dateutil/rrule.py", line 861, in _iter
    if until and res > until:
TypeError: can't compare offset-naive and offset-aware datetimes
```

EDIT: I monkey patched rrule.py after line 446: 
https://github.com/dateutil/dateutil/blob/1c8c90338a696c8c146c7eb6cc1bf52016912852/dateutil/rrule.py#L446
by adding
```
        if until:
            until = until.replace(tzinfo=self._tzinfo)
```

If you're interested I could make a PR, but maybe you have a better way in mind.

**Comment by jeinarsson** [comment]
Thanks for your work and answer.

The RFC actually says that UNTIL must be the same type as DTSTART, specifically (p.40)

> The value of the UNTIL rule part MUST have the same
> value type as the "DTSTART" property.  Furthermore, if the
> "DTSTART" property is specified as a date with local time, then
> the UNTIL rule part MUST also be specified as a date with local
> time.  If the "DTSTART" property is specified as a date with UTC
> time or a date with local time and time zone reference, then the
> UNTIL rule part MUST be specified as a date with UTC time.  In the
> case of the "STANDARD" and "DAYLIGHT" sub-components the UNTIL
> rule part MUST always be specified as a date with UTC time.  If
> specified as a DATE-TIME value, then it MUST be specified in a UTC
> time format.

So in my particular case above, I should fix my problems outside dateutil, because my call was not in line with RFC 5545.

There is, however, some type mangling going on (`date`s are upgraded to non-aware `datetime`s) that cause minor confusion (say when trying to give limits to `between(..)` one has to be aware of these types, because date cannot be compared to datetime cannot be compared to tz-aware datetime..). I have not yet reached a conclusion if this confusion is my problem or more rrule in general.

**Comment by pganssle** [comment]
CC @ryanpetrello This is one situation where following the RFC exactly seems like it *could* break things, particularly since we don't give any access to the individual tokens in the rrule string, so people won't have any way to fix broken strings emitted by some other service if we start becoming stricter about this.

One thing that's unfortunate here is that I have no real usage data here. The best way to find out if this would break things is to just make the change and see if people start complaining, but that's not really ideal.

**Comment by ryanpetrello** [comment]
@pganssle makes sense; it we _wanted_ to give this a shot, I _think_ https://github.com/dateutil/dateutil/pull/634 would accomplish it.

**Comment by pganssle** [comment]
Actually, it occurs to me that it's not like this works *now*, right? Anyone encountering this will actually just get a confusing error message, not something that works, so now might be the best time to make this change.

**Comment by ryanpetrello** [comment]
Yea, right now it's _broken_.  With my PR, it still fails, but at least it tells you _why_ it couldn't parse the rrule.

**Comment by pganssle** [comment]
Fixed in #634
